### ISSUE TITLE
Stack overflow due to unbounded getFauxShuffleMask() recursion

### ISSUE THREAD

**Issue by c0d1f1ed** [llvm/llvm-project#52960]
`getFauxShuffleMask()` takes a `Depth` parameter but doesn't actually use it to limit recursion depth.

I suggest adding the following:
```
  if (Depth > SelectionDAG::MaxRecursionDepth)
    return alse;
```

**Comment by RKSimon** [comment]
@c0d1f1ed Do you have a repro test case please?

**Comment by c0d1f1ed** [comment]
The issue was detected while running Vulkan graphics tests for Android apps on Chrome OS, on top of the SwiftShader CPU-based driver. The latter uses LLVM as a JIT-compiler. The stack overflow only happens in Debug builds. Teasing out a useful standalone repro could prove a bit challenging.

For what it's worth, our downstream patch has been confirmed to address the issue: https://swiftshader-review.googlesource.com/c/SwiftShader/+/60148.

**Comment by RKSimon** [comment]
OK cheers - by inspection it makes sense - although we're better off with the limit in getTargetShuffleInputs and not the getFauxShuffleMask helper.

An annoyance is we see some test regressions with the change, so there's a little yak shaving to do - I'll get on it.
